712 - `shinytest2` for `tm_variable_browser`

averissimo opened the pull request:
# Pull Request

<!--- Replace `#nnn` with your issue link for reference. -->

Part of  #712 

#### Changes description

- prefixes `ggplot2` functions calls
- "content is displayed correctly" test
  - Ensures that all datasets are present in left table
  - Ensures that table below the plot has correct first column title ("Statistic" for numeric variables and "levels" for categorical)
- "main output interactivity doesn't show errors" test
  - Tests all buttons and input sliders

#### Notes:

- Click function on javascript elements with callbacks doesn't trigger them
  - Clicking on variable table row does not select them

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+  )
+
+  if (!is.infinite(categorical_index)) {
+    app_driver$set_module_input(
```
```suggestion
    app_driver$set_active_module_input(
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+      categorical_index,
+      allow_no_input_binding_ = TRUE
+    )
+    app_driver$set_module_input(
```
```suggestion
    app_driver$set_active_module_input(
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+    1
+  ) ## Added to UI
+
+  app_driver$set_module_input("outlier_definition_slider", 2)
```
```suggestion
  app_driver$set_active_module_input("outlier_definition_slider", 2)
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+  # Test changing plot settings
+  testthat::expect_false(app_driver$is_visible(selector_ns("ggplot_theme-selectized")))
+
+  app_driver$set_module_input("ggplot_theme", "bw")
```
```suggestion
  app_driver$set_active_module_input("ggplot_theme", "bw")
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+
+  app_driver$set_module_input("ggplot_theme", "bw")
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("ggplot_theme", "light")
```
```suggestion
  app_driver$set_active_module_input("ggplot_theme", "light")
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("ggplot_theme", "light")
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("ggplot_theme", "dark")
```
```suggestion
  app_driver$set_active_module_input("ggplot_theme", "dark")
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+  app_driver$set_module_input("ggplot_theme", "dark")
+  app_driver$expect_no_validation_error()
+
+  app_driver$set_module_input("font_size", "8")
```
```suggestion
  app_driver$set_active_module_input("font_size", "8")
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+
+  app_driver$set_module_input("font_size", "8")
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("font_size", "20")
```
```suggestion
  app_driver$set_active_module_input("font_size", "20")
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("font_size", "20")
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("font_size", "15")
```
```suggestion
  app_driver$set_active_module_input("font_size", "15")
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+  app_driver$set_module_input("font_size", "15")
+  app_driver$expect_no_validation_error()
+
+  app_driver$set_module_input("label_rotation", "25")
```
```suggestion
  app_driver$set_active_module_input("label_rotation", "25")
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+
+  app_driver$set_module_input("label_rotation", "25")
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("label_rotation", "0")
```
```suggestion
  app_driver$set_active_module_input("label_rotation", "0")
```

m7pr commented on tests/testthat/test-shinytest2-tm_variable_browser.R:
```diff
@@ -0,0 +1,134 @@
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("label_rotation", "0")
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("label_rotation", "90")
```
```suggestion
  app_driver$set_active_module_input("label_rotation", "90")
```

m7pr reviewed (changes requested):
rename `set_module_input` to `set_active_module_input`

kartikeyakirar reviewed (changes requested):
@averissimo test passes with warning. 
![image](https://github.com/insightsengineering/teal.modules.general/assets/6700955/808b2515-fcdf-4418-bf20-7c91d9f505d2)

Apart from that few additional request:

1: We don't need to call app_driver$expect_no_shiny_error() after every initialization.
2: Could you add tests to check the visibility of plots using app_driver$is_visible()?
2. is is possible to add test for checkbox (teat variable as factor)
<img width="689" alt="image" src="https://github.com/insightsengineering/teal.modules.general/assets/6700955/0a380c9e-6025-4c12-bc27-983d71d63769">

averissimo commented:
Thanks for the comments @kartikeyakirar.

I implemented them on the two new commits and split some of the tests that were too long / big scope.

> 1: We don't need to call app_driver$expect_no_shiny_error() after every initialization.

Does it seem fine as it is, or did you also mean to remove some of them after changing the inputs?

kartikeyakirar commented:
> Does it seem fine as it is, or did you also mean to remove some of them after changing the inputs?

I'm fine with either option, but since this additional test has already been checked and Marcin has removed it from another PR, let's keep it consistent.

kartikeyakirar reviewed (approved):
Great Work
